### PR TITLE
Dev 114 bmrs eic dukes

### PR DESCRIPTION
**Pull request by msff2**
@mdhillman John and I addressed your following comments.

a) Update the README file (TheWorldAvatar\UK_Power_Generation_Units_BMRS\Python\Code\ukpowergeneration\README.txt) of the UKPowerGeneration project with a generic description.

b) Describe the data source used in the project in the same README file.

c) Provide a screenshot of the visualisation interface in the README file to help future users understand if their generated output looks alike.

### CODE REVIEW

**Comment by mdhillman** [comment]
I doubt I will have to look at this before I leave CMCL, as such I'm adding @gpeb2 and @Ushcode (at least one of them should take a look).

I can already say though that I think a better file location for the project can be found, I'm not a fan of adding new top-level directories to the repository. Perhaps the `/web` directory is a better home?

**Comment by sm453** [comment]
How does this PR relate to dev-migrate-johns-projects-to-stack? Is that going to be discarded? Or will there be a separate PR for that?

**Comment by msff2** [comment]
> How does this PR relate to dev-migrate-johns-projects-to-stack? Is that going to be discarded? Or will there be a separate PR for that?

That is going to be discarded.

**Comment by msff2** [comment]
> I doubt I will have to look at this before I leave CMCL, as such I'm adding @gpeb2 and @Ushcode (at least one of them should take a look).
> 
> I can already say though that I think a better file location for the project can be found, I'm not a fan of adding new top-level directories to the repository. Perhaps the `/web` directory is a better home?

Good point. We will address this.

**Comment by sm453** [comment]
> > How does this PR relate to dev-migrate-johns-projects-to-stack? Is that going to be discarded? Or will there be a separate PR for that?
> 
> That is going to be discarded.

Just for the record: Having discussed it, a preferable action will be to merge dev-migrate-johns-projects-to-stack as part of a separate PR, once this PR is merged.

File: UK_Power_Generation_Units_BMRS/Python/Code/ukexportcurtailmentimporter/resources/ukexportcurtailmentimporter.properties [hunk 1]
```diff
@@ -0,0 +1,25 @@
+###--- Properties for Python script ---###
+
+# Output directory for JSON files containing retrieved data
+# In deployment under Docker, this should be /var/www/html/ukpowergeneration
+output.directory = ./visualisation/data
+
+###--- Properties for Digital Twin Visualisation Framework ---###
+
+# Mapbox API key (to be copied to "index.html" file)
+mapbox.apiKey = pk.eyJ1IjoibXNmZjIiLCJhIjoiY2t2bDRuZG9xMDg3dzJvbTh4NnNremgydyJ9.H5CctfdCnbxZDEt7R1w3yA
```
Comment (sm453) [inline UK_Power_Generation_Units_BMRS/Python/Code/ukexportcurtailmentimporter/resources/ukexportcurtailmentimporter.properties:10]: Please remove the key value.

File: UK_Power_Generation_Units_BMRS/Python/Code/ukpowergeneration/ScriptMapQuery/README-BMRS-API-Input-JA-#-ReadAll.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+## Description
+This document is a README for:
+> BMRS-API-Input-JA-#-ReadAll.py
+
+Three csv files are required. These are not version controlled, but can be found at: 
+> \Dropbox (Cambridge CARES)\CoMo shared\ja685\BMRS\Script-BMRS-API
+Their names are: 
+> Input-Template.csv
+> generator.tripple.csv
+> powerplanttripple.csv
```
Comment (sm453) [inline UK_Power_Generation_Units_BMRS/Python/Code/ukpowergeneration/ScriptMapQuery/README-BMRS-API-Input-JA-#-ReadAll.md:11]: Please update to:
`dropbox\CoMo shared\ja685\Modelling-Resources\BMRS\Script-BMRS-API`

File: UK_Power_Generation_Units_BMRS/Python/Code/ukpowergeneration/ScriptMapQuery/README-Mapping-Match-JA-#.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+## Description
+This document is a README for:
+ > Mapping-Match-JA-#.py
+
+Three csv files are required. These are not version controlled, but can be found at: 
+> \Dropbox (Cambridge CARES)\CoMo shared\ja685\BMRS\Script-BMRS-API
+Their names are: 
+> Input-Template.csv
+> generatortripple.csv
+> powerplanttripple.csv
```
Comment (sm453) [inline UK_Power_Generation_Units_BMRS/Python/Code/ukpowergeneration/ScriptMapQuery/README-Mapping-Match-JA-#.md:11]: Please update to:
`dropbox\CoMo shared\ja685\Modelling-Resources\BMRS\Script-BMRS-API`

File: UK_Power_Generation_Units_BMRS/Python/Code/ukpowergeneration/mapping/README-Mapping-Match-JA-#.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+## Note
+This is not called by the main functions, but is instead used manually to map, with the output being done ‘Input….csv’ (which must be on this level for the main file ‘Mapping_Match_JA_9.py’ to be run on it). 
+Manually checking the output is reccomended. 
+
+## Description
+This document is a README for:
+ > Mapping-Match-JA-#.py
+
+Three csv files are required. These are not version controlled, but can be found at: 
+> \Dropbox (Cambridge CARES)\CoMo shared\ja685\BMRS\Script-BMRS-API
```
Comment (sm453) [inline UK_Power_Generation_Units_BMRS/Python/Code/ukpowergeneration/mapping/README-Mapping-Match-JA-#.md:12]: Please update to:
`dropbox\CoMo shared\ja685\Modelling-Resources\BMRS\Script-BMRS-API`

File: UK_Power_Generation_Units_BMRS/Python/Code/ukpowergeneration/ScriptMapQuery/BMRS_API_Input_JA_7.py [hunk 5]
```diff
@@ -0,0 +1,526 @@
+#KEY: iwx6raw9m7nqq0f (this should not be in the Version Control).
```
Comment (sm453) [inline UK_Power_Generation_Units_BMRS/Python/Code/ukpowergeneration/ScriptMapQuery/BMRS_API_Input_JA_7.py:1]: Please remove this line.

File: UK_Power_Generation_Units_BMRS/Python/Code/ukpowergeneration/ScriptMapQuery/BMRS_API_Input_JA_7.py [hunk 6]
```diff
@@ -0,0 +1,526 @@
...
+    #The query function
+    Key = str(Key)
+    Year = str(Year)
+    Month = str(Month)
+    Day = str(Day)
+    Period = str(Period)
+    querystring = 'https://api.bmreports.com/BMRS/B1610/v2?APIKey='+Key+'&SettlementDate='+Year+'-'+Month+'-'+Day+'&Period='+Period+'&NGCBMUnitID=&ServiceType=xml'
+    
+    #Query All
+    #post_elexon_basic(url='https://api.bmreports.com/BMRS/B1610/v2?APIKey=iwx6raw9m7nqq0f&SettlementDate=2021-01-01&Period=1&NGCBMUnitID=&ServiceType=xml',) #All
```
Comment (sm453) [inline UK_Power_Generation_Units_BMRS/Python/Code/ukpowergeneration/ScriptMapQuery/BMRS_API_Input_JA_7.py:297]: Please remove the API key value from lines like this (there are a few in this file). I would not remove the lines altogether - they might still be useful.

File: UK_Power_Generation_Units_BMRS/Python/Code/ukpowergeneration/resources/ukpowergeneration.properties [hunk 7]
```diff
@@ -0,0 +1,25 @@
+###--- Properties for Python script ---###
+
+# Output directory for JSON files containing retrieved data
+# In deployment under Docker, this should be /var/www/html/ukpowergeneration
+output.directory = ./visualisation/data
+
+###--- Properties for Digital Twin Visualisation Framework ---###
+
+# Mapbox API key (to be copied to "index.html" file)
+mapbox.apiKey = pk.eyJ1IjoibXNmZjIiLCJhIjoiY2t2bDRuZG9xMDg3dzJvbTh4NnNremgydyJ9.H5CctfdCnbxZDEt7R1w3yA
```
Comment (sm453) [inline UK_Power_Generation_Units_BMRS/Python/Code/ukpowergeneration/resources/ukpowergeneration.properties:10]: Please remove the key value (not the whole line).

File: UK_Power_Generation_Units_BMRS/Python/Code/ukpowergeneration/visualisation/index.html [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+	<script>
+		// This ensures that the "manager" variable can be accessed from 
+		// HTMLElements for interaction events.
+		let manager = null;
+
+		// Enter your MapBox API here.
+		// This will be visible by inspecting the webpage anyway (either by looking at the source code,
+		// or inspecting the network packets). If it needs protection, limit it's usage to certain source
+		// domains via your MapBox account settings.
+		let mapboxAPI = "pk.eyJ1IjoibXNmZjIiLCJhIjoiY2t2bDRuZG9xMDg3dzJvbTh4NnNremgydyJ9.H5CctfdCnbxZDEt7R1w3yA";
```
Comment (sm453) [inline UK_Power_Generation_Units_BMRS/Python/Code/ukpowergeneration/visualisation/index.html:82]: Please remove the key value (not the whole line).

File: UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Marginal-Seller/BMRS-API-JA-MarginalSeller.py [hunk 9]
```diff
@@ -0,0 +1,210 @@
+#KEY: iwx6raw9m7nqq0f
```
Comment (sm453) [inline UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Marginal-Seller/BMRS-API-JA-MarginalSeller.py:1]: Please remove this line.

File: UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Marginal-Seller/BMRS-API-JA-MarginalSeller.py [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+    #print(type(pformat(content)))
+
+    #print('===Finished===')
+    
+    return pformat(content)
+
+
+def run_price_query(YYYY, MM, DD, PP):
+    #Query All
+    makeurl = 'https://api.bmreports.com/BMRS/B1770/v1?APIKey=iwx6raw9m7nqq0f&SettlementDate='+YYYY+'-'+MM+'-'+DD+'&Period='+PP+'&ServiceType=xml'
```
Comment (sm453) [inline UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Marginal-Seller/BMRS-API-JA-MarginalSeller.py:57]: Please remove the API key value from lines like this (there are a few in this file). Don't remove the lines as a whole.

File: UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Per-Wind-Unit-Curtailment/BMRS-API-JA-PerGenerationUnitCurtailment-Bulk-2.py [hunk 11]
```diff
@@ -0,0 +1,154 @@
+#KEY: iwx6raw9m7nqq0f
```
Comment (sm453) [inline UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Per-Wind-Unit-Curtailment/BMRS-API-JA-PerGenerationUnitCurtailment-Bulk-2.py:1]: Please remove this line.

File: UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Per-Wind-Unit-Curtailment/BMRS-API-JA-PerGenerationUnitCurtailment-Bulk-2.py [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+    
+    return pformat(content)
+
+
+def query_curtailment(YYYY, MM, DD, PP):
+    #NOTE: IF YOU WANT PRICES IN THE FUTURE, then I reccomend duplicating this function, but going by bidPrice (or something else like that once you've checked what is needed) instead of bidVolume.
+    #Keep in mind, however, that there's a but and sell side to keep track of (think the buy side might use something like offerPrice and offerVolume, rather than bidPrice and bidVolume), so you'd need to consider this change too.
+    #Currently the bidding is not used here, but rather, as this is for curtailment, sell side volumes are considered.
+    #This process is also specifically used for renewables. 
+    queryoutput = post_elexon(url='https://api.bmreports.com/BMRS/DETSYSPRICES/v1?APIKey=iwx6raw9m7nqq0f&SettlementDate='+YYYY+'-'+MM+'-'+DD+'&SettlementPeriod='+PP+'&ServiceType=xml',) #All
```
Comment (sm453) [inline UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Per-Wind-Unit-Curtailment/BMRS-API-JA-PerGenerationUnitCurtailment-Bulk-2.py:60]: Please remove the API key value from lines like this. Don't remove the lines as a whole.

File: UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Per-Wind-Unit-Export/BMRS-API-JA-PerGenerationUnit-Bulk-2.py [hunk 13]
```diff
@@ -0,0 +1,154 @@
+#KEY: iwx6raw9m7nqq0f
```
Comment (sm453) [inline UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Per-Wind-Unit-Export/BMRS-API-JA-PerGenerationUnit-Bulk-2.py:1]: Please remove this line.

File: UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Per-Wind-Unit-Export/BMRS-API-JA-PerGenerationUnit-Bulk-2.py [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+    #print(pformat(content))
+    #print(type(pformat(content)))
+
+    #print('===Finished===')
+    
+    return pformat(content)
+
+
+def query_units(YYYY, MM, DD):
+    queryoutput = post_elexon(url='https://api.bmreports.com/BMRS/B1610/v2?APIKey=iwx6raw9m7nqq0f&SettlementDate='+YYYY+'-'+MM+'-'+DD+'&Period=*&NGCBMUnitID=&ServiceType=xml',) #All
```
Comment (sm453) [inline UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Per-Wind-Unit-Export/BMRS-API-JA-PerGenerationUnit-Bulk-2.py:56]: Please remove the API key value from lines like this. Don't remove the lines as a whole.

File: UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Price-And-Aggregate-By-Type/BMRS-API-PRICE-TYPE-3-FullTime.py [hunk 15]
```diff
@@ -0,0 +1,221 @@
+#KEY: iwx6raw9m7nqq0f
```
Comment (sm453) [inline UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Price-And-Aggregate-By-Type/BMRS-API-PRICE-TYPE-3-FullTime.py:1]: Please remove this line.

File: UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Price-And-Aggregate-By-Type/BMRS-API-PRICE-TYPE-3-FullTime.py [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+        print(pformat(content))
+        print(type(pformat(content)))
+
+        print('===Finished===')
+    return pformat(content)
+
+
+def run_query(YYYY, MM, DD, PP):
+    #Query All
+    makeurl = 'https://api.bmreports.com/BMRS/B1770/v1?APIKey=iwx6raw9m7nqq0f&SettlementDate='+YYYY+'-'+MM+'-'+DD+'&Period='+PP+'&ServiceType=xml'
```
Comment (sm453) [inline UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Price-And-Aggregate-By-Type/BMRS-API-PRICE-TYPE-3-FullTime.py:59]: Please remove the API key from lines like this (there are a few in this file). Don't remove the lines as a whole.

**Review by sm453 — CHANGES_REQUESTED** [review]
I agree with @mdhillman that this should not be at root level. I'd prefer it in `Agents/`, though (even if it isn't one...).

File: UK_Power_Generation_Units_BMRS/Python/Code/ukpowergeneration/visualisation/index.html [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+	<script src='./js/manager.js'></script>
+	<script src='./js/data_registry.js'></script>
+	<script src='./js/source_handler.js'></script>
+	<script src='./js/layer_handler.js'></script>
+	<script src='./js/control_handler.js'></script>
+	<script src='./js/panel_handler.js'></script>
+	<script src='./js/interaction_handler.js'></script>
+	<script src='./js/timeseries_handler.js'></script>
+	<script src='./js/icon_handler.js'></script>
+	<script src='./js/jsonview.bundle.js'></script>
```
Comment (Ushcode) [inline UK_Power_Generation_Units_BMRS/Python/Code/ukpowergeneration/visualisation/index.html:43]:
```suggestion
	        <script src="./twa-vf/twa-vf.min.js" charset="utf-8"></script>

```

remove this and replace with twavf image

**Comment by Ushcode** [comment]
image


viz looks like this

**Comment by Ushcode** [comment]
It would be better to spin this up as part of a stack and use the latest `twa-vf` rather than copying in the `DTVF` files to a different location. They shouldn't be duplicated and are now two years old

**Comment by Ushcode** [comment]
I've moved the `UK_Power_Generation_Units_BMRS` directory into `/Agents`

**Comment by Ushcode** [comment]
I don't have access to `Dropbox (Cambridge CARES)`. May need to commit to the repo to test this

**Comment by sm453** [comment]
> I don't have access to `Dropbox (Cambridge CARES)`. May need to commit to the repo to test this

Just for the record, please _never ever_ commit data to a source code repository. (Ok, exceptions exist, but I don't believe this is one.)

**Comment by Ushcode** [comment]
> > I don't have access to `Dropbox (Cambridge CARES)`. May need to commit to the repo to test this
> 
> Just for the record, please _never ever_ commit data to a source code repository. (Ok, exceptions exist, but I don't believe this is one.)

'Input-Template-Auto.csv' and 'LocalOnlyBMRSKey.txt' do not sound like data to me. It does not make sense to me to merge code to the public TWA repository that depends on files in a private dropbox folder

File: UK_Power_Generation_Units_BMRS/Python/Code/ukexportcurtailmentimporter/README-ExcelToTimeSeries-TimeStampFunctions.txt [hunk 18]
```diff
@@ -0,0 +1 @@
+Further functions of assistance from "\Dropbox (Cambridge CARES)\CoMo shared\ja685\BMRS\Wind-and-ESS\ESSPythonScript\Supplementary".
```
[the remark this answers is not in the document]
Comment (msff2) [reply UK_Power_Generation_Units_BMRS/Python/Code/ukexportcurtailmentimporter/README-ExcelToTimeSeries-TimeStampFunctions.txt]: Done.

File: UK_Power_Generation_Units_BMRS/Python/Code/ukpowergeneration/ScriptMapQuery/README-Mapping-Match-JA-#.md (hunk 3 above)
Comment (msff2) [reply UK_Power_Generation_Units_BMRS/Python/Code/ukpowergeneration/ScriptMapQuery/README-Mapping-Match-JA-#.md]: Done.

File: UK_Power_Generation_Units_BMRS/Python/Code/ukpowergeneration/mapping/README-Mapping-Match-JA-#.md (hunk 4 above)
Comment (msff2) [reply UK_Power_Generation_Units_BMRS/Python/Code/ukpowergeneration/mapping/README-Mapping-Match-JA-#.md]: Done.

File: UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Per-Wind-Unit-Curtailment/BMRS-API-JA-PerGenerationUnitCurtailment-Bulk-2.py (hunk 12 above)
Comment (msff2) [reply UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Per-Wind-Unit-Curtailment/BMRS-API-JA-PerGenerationUnitCurtailment-Bulk-2.py]: Done.

File: UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Per-Wind-Unit-Export/BMRS-API-JA-PerGenerationUnit-Bulk-2.py (hunk 13 above)
Comment (msff2) [reply UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Per-Wind-Unit-Export/BMRS-API-JA-PerGenerationUnit-Bulk-2.py]: Done.

File: UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Per-Wind-Unit-Export/BMRS-API-JA-PerGenerationUnit-Bulk-2.py (hunk 14 above)
Comment (msff2) [reply UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Per-Wind-Unit-Export/BMRS-API-JA-PerGenerationUnit-Bulk-2.py]: Done.

File: UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Price-And-Aggregate-By-Type/BMRS-API-PRICE-TYPE-3-FullTime.py (hunk 15 above)
Comment (msff2) [reply UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Price-And-Aggregate-By-Type/BMRS-API-PRICE-TYPE-3-FullTime.py]: Done.

File: UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Price-And-Aggregate-By-Type/BMRS-API-PRICE-TYPE-3-FullTime.py (hunk 16 above)
Comment (msff2) [reply UK_Power_Generation_Units_BMRS/Python/Code/ukpowersupply/Script-BMRS-API-Price-And-Aggregate-By-Type/BMRS-API-PRICE-TYPE-3-FullTime.py]: Done.

**Review by sm453 — APPROVED** [review]

**Pull request merged**